types/wlr_output_power_management_v1.c

```diff
@@ -0,0 +1,244 @@
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwlr_output_power_v1_interface,
+		&output_power_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void output_power_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_output_power_v1 *output_power =
+		output_power_from_resource(resource);
+	output_power_destroy(output_power);
+}
+
+static void output_power_handle_output_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_output_power_v1 *output_power =
+		wl_container_of(listener, output_power, output_destroy_listener);
+	output_power_destroy(output_power);
+}
+
+static void wlr_output_power_v1_send_mode(struct wlr_output_power_v1 *output_power) {
+	enum zwlr_output_power_v1_mode mode;
+
+	mode = output_power->output->enabled ? ZWLR_OUTPUT_POWER_V1_MODE_ON :
+		ZWLR_OUTPUT_POWER_V1_MODE_OFF;
+	zwlr_output_power_v1_send_mode(output_power->resource, mode);
+}
+
+static void output_power_handle_output_enable(struct wl_listener *listener,
+		void *data) {
+	struct wlr_output_power_v1 *output_power =
+		wl_container_of(listener, output_power, output_enable_listener);
+	wlr_output_power_v1_send_mode(output_power);
+}
+
+static void output_power_handle_set_mode(struct wl_client *client,
+		struct wl_resource *output_power_resource,
+		enum zwlr_output_power_v1_mode mode) {
+	struct wlr_output_power_v1 *output_power =
+		output_power_from_resource(output_power_resource);
+	if (output_power == NULL) {
+		return;
+	}
+
+	switch (mode) {
+	case ZWLR_OUTPUT_POWER_V1_MODE_OFF:
+	case ZWLR_OUTPUT_POWER_V1_MODE_ON:
+		break;
+	default:
+		wlr_log(WLR_ERROR, "Invalid power mode %d", mode);
+		zwlr_output_power_v1_send_failed(output_power->resource);
+	}
+
+	bool ok = wlr_output_enable(output_power->output, mode);
+	if (!ok) {
+		zwlr_output_power_v1_send_failed(output_power->resource);
+	}
+}
+
+static const struct zwlr_output_power_v1_interface output_power_impl = {
+	.destroy = output_power_handle_destroy,
+	.set_mode = output_power_handle_set_mode,
+};
+
+static const struct zwlr_output_power_manager_v1_interface
+	output_power_manager_impl;
+
+static struct wlr_output_power_manager_v1 *output_power_manager_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwlr_output_power_manager_v1_interface, &output_power_manager_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void output_power_manager_get_output_power(struct wl_client *client,
+		struct wl_resource *manager_resource, uint32_t id,
+		struct wl_resource *output_resource) {
+	struct wlr_output_power_manager_v1 *manager =
+		output_power_manager_from_resource(manager_resource);
+	struct wlr_output *output = wlr_output_from_resource(output_resource);
+
+	struct wlr_output_power_v1 *output_power =
+		calloc(1, sizeof(struct wlr_output_power_v1));
+	if (output_power == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	output_power->output = output;
```

Comment:
`output` can be NULL at this point, in case it's inert.

Reply:
modified to send
```
			zwlr_output_power_v1_send_failed(mgmt->resource);
```

too in this case. should we mention this in the protocol?